rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,391 @@
+flag.
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate    # <== Now a source template
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         name: $(workload.metadata.name)$-linter      # <== Refers to workload
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$      # <== Refers to workload
+            - name: revision
+              value: $(workload.spec.source.git.revision)$ # <== Refers to workload
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   lifecycle: immutable                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the immutable flag, rather than
+creating/updating the templated object it will create/update a Runnable and
+ensure that a matching ClusterRunTemplate exists.
+
+What is meant by "ensure that a matching ClusterRunTemplate exists"? For any
+given supply chain template (a cluster scoped object) there will only need
+to be one ClusterRunTemplate. All of the N Runnables created by N workloads
+will reference this one ClusterRunTemplate. Only if the supply chain
+template is updated will this ClusterRunTemplate object need updating.
+
+## Inputs
+The templated object will have two types of fields: hard-coded and templated.
+Hard coded fields can simply be copied directly into the ClusterRunTemplate.
+Templated fields will be translated. Rather than reading from the specified
+location, these will read from the runnable's inputs field. This can be done
+simply by prepending the template paths with runnable.spec.inputs
+
+Example:
+```yaml
+kind: ClusterSourceTemplate    # <== Now a source template
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$
+```
+
+Will translate to a ClusterRunTemplate with:
+```yaml
+kind: ClusterRunTemplate
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(runnable.spec.inputs.workload.spec.source.git.url)$
+```
+
+and a Runnable with
+```yaml
+kind: Runnable
+spec:
+   inputs:
+      workload.spec.source.git.url: *some-value-from-workload.spec.source.git.url*
+```
+
+## Names
+The runnable will be created with the same name assigned to the templated
+object with the object kind appended.
```

Comment:
pretty interesting - I _think_ this has been discussed elsewhere, but, wdyt of using `generateName-`? that way we can be sure that the name we got is not conflicting with anything in that namespace

Reply:
I'm not opposed to using generate name for the runnable. I expect it will require slightly more processing by our workload, as we would need to list all the runnables in the namespace and filter for the one that we want.